Home screen/home screen cards

ofekcMoveo opened the pull request:
in this branch I created and implemented the delegate pattern on the tableView and tableViewCell so when a button is pressed in the article card (cell) it will be passed to the table itself and there I will handle the event (in the mean time just printing the title of the article). Also I designed the table's and cell's UI

ofekcMoveo committed:
created a table view, and started working on tavbeViewCell (added an image), created the Article and Articles models with mock data

ofekcMoveo committed:
continue design articleCell (added button, titles, and tags labels), open ArticleCellProtocol

ofekcMoveo committed:
right arrow icon add to dispatch button + add to favorites button added

ofekcMoveo committed:
ui fixes

ofekcMoveo committed:
create and implement delegate pattern to the table cells, fixing ui of homepage - background of the table, spaces between cells, border of cells..

ofekcMoveo committed:
fix the height of the UITableview

noamkuMoDev commented on Dispatcher/Controllers/HomeViewController.swift:
```diff
@@ -27,3 +38,37 @@ class HomeViewController: UIViewController {
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: ArticleCell.articleCellIdentifier, for: indexPath) as! ArticleCell
+        
+        cell.dateLabel.text = articlesToDisplay[indexPath.row].Date
```
It will be more readable if you save the articlesToDisplay[indexPath.row] in a variable at the beginning of the function

noamkuMoDev commented on Dispatcher/Models/ArticleTableViewDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+protocol ArticleTableViewDelegate {
```
The delegate is more about the cell of the table than about the article, thats why the methods inside it are about navigation and adding to favorites (and not something article-oriented).
Therefore, a better name for the delegate would be "NewsCellDelegate".
Also, we would prefer to place the protocol inside the ArticleCell.swift file instead of giving it an entire file of its own.

noamkuMoDev commented on Dispatcher/Models/ArticleTableViewDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+
+protocol ArticleTableViewDelegate {
+    func navigateToArticle(_ articleTitle: String)
+    func addArticleToFavorites(_ articleTitle: String)
```
Names of methods inside delegate protocols usually (according to the convention) are given names like "actionButtonDidPress" or "backButtonTapped". This way it describes the action that happened, and the code that should be executed (e.g. performing a navigation) is done inside the implementation itself.

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+    @IBOutlet weak var subTitleLabel: UILabel!
+    @IBOutlet weak var dispatchButton: UIButton!
+    @IBOutlet weak var moreTagsLabel: UILabel!
+    @IBOutlet weak var addToFavoritesButton: UIButton!
```
The button can either add to favorites or remove from favorites. Therefore its name is somewhat misleading and should be changed

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+    @IBOutlet weak var moreTagsLabel: UILabel!
+    @IBOutlet weak var addToFavoritesButton: UIButton!
+    
+    var delegate: ArticleTableViewDelegate? = nil
```
FYI you don't need to give it a default value of nil by yourself - the moment you define it as optional and don't give it an initial value, then it automatically receives the default value of nil

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+        delegate?.navigateToArticle(self.titleLabel.text!)
+    }
+    
+    @IBAction func addToFavoritesButtonPressed(_ sender: UIButton) {
```
Same thing here about the name of the function, as pressing the same button can result in two different outcomes - removing from favorites as well as adding to favorites

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+    
+    @IBAction func navigateButtonPressed(_ sender: UIButton) {
+        delegate?.navigateToArticle(self.titleLabel.text!)
```
It's risky to use ! as it can lead to crushes if for some reason we don't actually have a value inside our text label. It is better to use ?? and provide a default value

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+        self.layer.cornerRadius = 20
```
If you perform more than a couple of actions on the same element, it is better to extract them into a dedicated function and call this function wherever needed

noamkuMoDev commented on Dispatcher/Views/Base.lproj/Main.storyboard:
```diff
@@ -146,28 +138,45 @@
+                                <rect key="frame" x="0.0" y="152" width="393" height="700"/>
+                                <color key="backgroundColor" name="screenBackgroundColor"/>
+                                <constraints>
+                                    <constraint firstAttribute="height" constant="700" id="LFR-iF-fdQ"/>
```
You can't give a fixed height to a tableView - it might look good on the device that you are building on, but it will be too big or too small on another device. Find different constraints to achieve the design.

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,62 @@
+        contentView.layoutMargins.left = 15
+        contentView.layoutMargins.right = 15
+
+        tagLabel.layer.cornerRadius = 10
```
Make sure to add text colors, because on older iPhones the default text color can suddenly be different than you expect which results in bugs (e.g. on my iPhone 7 all the texts are white and therefore none of your cells is readable)

noamkuMoDev replied:
Still relevant, add the text colors hardcoded and not only through the xcode right pane. When I run the app on my device it looks like this:
![image](https://user-images.githubusercontent.com/100192030/234014072-18969496-249c-476a-8829-3bb607a86588.png)

ofekcMoveo committed:
favoritesButton name changed and also the relevant IBAction name
working with optionals fixed
extract some actions from awaikFromNib to a different function and call here from awaikFromNib
delegate protocol name changed and the protocol moved to the ArticleCell file

ofekcMoveo committed:
fix text colors and table view constraints

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,76 @@
+
+    
+    @IBAction func navigateButtonPressed(_ sender: UIButton) {
+        delegate?.navigateButtonPressed(self.titleLabel.text ?? "")
```
Unless we are inside a closure (and some other very specific cases), we don't need to put the self before referring to a local variable. So if not needed, let's avoid it so we have shorter code lines

noamkuMoDev commented on Dispatcher/Models/Article.swift:
```diff
@@ -0,0 +1,17 @@
+    let subTitle: String
+    let auther: String
+    let tags: [String]
+    let Date: String
```
properties should always start with a small letter (date and not Date)

noamkuMoDev commented on Dispatcher/Controllers/HomeViewController.swift:
```diff
@@ -26,4 +36,40 @@ class HomeViewController: UIViewController {
+        cell.titleLabel.text = currentArticle.title
+        cell.subTitleLabel.text = currentArticle.subTitle
+        cell.tagLabel.text = currentArticle.tags.first
+        cell.moreTagsLabel.text  = "+ \(currentArticle.tags.count - 1)"
```
If we have only 1 tag in our array, this will display a "+0" which is bad UX

noamkuMoDev commented on Dispatcher/Views/Base.lproj/Main.storyboard:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
You have some constraints issues in the HomePage in regards to the "Top Headlines In Israel"

<img width="436" alt="Screen Shot 2023-04-24 at 16 46 47" src="https://user-images.githubusercontent.com/100192030/234015634-64878a2f-e2c7-4d6c-968f-28dbe283a66e.png">

ofekcMoveo committed:
deleted  "Top Headlines in Israel" label and created a table headline instead
fixed big letter in Article's date property
added hardcoded colors to labels in ArticleCell
fixed tags and moreTags labels in ArticleCell in case of just one tag

noamkuMoDev commented on Dispatcher/Controllers/HomeViewController.swift:
```diff
@@ -26,4 +36,66 @@ class HomeViewController: UIViewController {
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: ArticleCell.articleCellIdentifier, for: indexPath) as! ArticleCell
```
Using as! is risky because if for some reason we can't make the forced downcast then our application will crash.
It's better practice to use "as?" and alternatively return a regular UITableCell because we prefer the user to see an empty cell rather than having the application crash.

noamkuMoDev commented on Dispatcher/Controllers/HomeViewController.swift:
```diff
@@ -26,4 +36,66 @@ class HomeViewController: UIViewController {
+}
+
+extension HomeViewController: ArticleCellDelegate, UITableViewDelegate {
+    func navigateButtonPressed(_ articleTitle: String) {
```
Thinking forward, we wouldn't want to perform actions based on the title of the article - but according to an ID of the article as it is more unique.
Add this property to your Article model and pass it in the delegate instead of the articleTitle

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,92 @@
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+        makeCardShape()
```
Let's give the function a more meaningful name so it's easier to understand what's inside it.
Since all of the code lines inside are about corner radius (except for the last one) let's call it "setCornerRadiusForCellElements" or something like that (and put the last code line back in the awakefromNib)

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,92 @@
+        
+    }
+    
+    private func setUIobjectsColor() {
```
Let's give the function a more meaningful name so it's easier to understand what's inside it (clean code rules).
Since all the code inside is about textColor - we can call it something like "setTextColorForCellElements"

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,92 @@
+        setUIobjectsShape()
+    }
+    
+    private func setUIobjectsShape() {
```
Let's give the function a more meaningful name so it's easier to understand what's inside it.
Since all of the code lines inside are about corner radius (except for the last one) let's call it "setCornerRadiusForCellElements" or something like that (and put the last code line back in the awakefromNib)

noamkuMoDev commented on Dispatcher/Views/ArticleCell.swift:
```diff
@@ -0,0 +1,92 @@
+        dispatchButton.setTitleColor(.white, for: .normal)
+    }
+    
+    private func makeCardShape() {
```
Let's give the function a more meaningful name so it's easier to understand what's inside it (clean code rules).
We can name it something like "setCellMarginsAndBorder"

ofekcMoveo committed:
change ArticleCell functions names
add id to an article
change downcasting in cellForRowAt func
